### PR TITLE
Improve speed of appending to large files

### LINKED ISSUE #102: Make appending efficient
The Linux 9p client writes large files (e.g. the Moby iso) 8KB at a time. Datakit handles each write by allocating a new buffer, copying the old version into it, and adding the new 8KB. This gets slow quickly ;-)

Change the type for files in `Ivfs_tree` from

      | Blob of Cstruct.t

to

      | Blob of Blob.t

and make a suitable `Blob` module for efficient non-destructive updates. e.g. copy-on-write, or just use a list of blocks in reverse order (and concat them all together at the end).

### PR DESCRIPTION
**Pull request by talex5**
Replace Cstruct.t with Ivfs_blob.t, which allows fast appending.

In a simple benchmark, writing a 102400 KB file in 800 KB chunks went from taking 436 to 2.8 seconds.

Fixes #102.

Signed-off-by: Thomas Leonard <thomas.leonard@docker.com>

### CODE REVIEW

**Comment by avsm** [comment]
Nice speedup!

File: src/ivfs/ivfs_blob.ml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  | 0L -> Ok empty
+  | new_len ->
+    let new_len = Int64.to_int new_len in
+    let extra = new_len - len old in
+    if extra = 0 then Ok old
+    else if extra < 0 then (
+      Ok (of_ro_cstruct (Cstruct.sub (to_ro_cstruct old) 0 new_len))
+    ) else (
+      let padding = Cstruct.create extra in
+      Cstruct.memset padding 0; (* Won't be needed with Cstruct >= 2.2 *)
```
Comment (avsm) [inline src/ivfs/ivfs_blob.ml]: I can release cstruct 2.2 now and then we can remove this and make that a hard constraint... is anyone else blocking cstruct 2.2?

File: src/ivfs/ivfs_blob.mli [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+val of_ro_cstruct: Cstruct.t -> t
+(** [of_ro_cstruct c] is a blob containing the data [c].
+    Note that we take a reference to [c] rather than copying, so [c] MUST NOT be modified
+    after this call. [c] may also be shared with modified versions of the resulting blob. *)
+
+val to_ro_cstruct: t -> Cstruct.t
+(** [to_ro_cstruct t] is a read-only Cstruct with the same data as [t].
+    DO NOT modify this - it may corrupt the blob or other blobs sharing data
+    with this one if you do. *)
```
Comment (avsm) [inline src/ivfs/ivfs_blob.mli]: Worth an feature request issue on Cstruct for properly typed RO buffers in a future revision?

Comment (talex5) [inline src/ivfs/ivfs_blob.mli]: Would be nice. I've added it to the long-standing phantom-type issue.

**Comment by samoht** [comment]
LGTM

**Comment by avsm** [comment]
LGTM

**Pull request merged**